### PR TITLE
Break up abstract connection pool

### PR DESCRIPTION
**Pull request by gmcgibbon**
### Summary

This is borderline a cosmetic change, but I was reading this file recently and was surprised by how big it was. The connection handler is big enough for it to be a separate file. The pool queues and reaper are also big enough to be separated I think, but that might be more of my preference than a necessity.

With this change, I believe the searchability of these public API classes improves enough for it to be worth it. What do you think?

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters.rb [hunk 1]
```diff
@@ -26,6 +26,10 @@ module ConnectionAdapters
     end
 
     autoload_at "active_record/connection_adapters/abstract/connection_pool" do
+      autoload :ConnectionPool
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters.rb:29]: ConnectionPool is already on line 41. It seems you only need to move `ConnectionHandler` to that block.

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/connection_adapters.rb (hunk 1 above)
Comment (gmcgibbon) [reply activerecord/lib/active_record/connection_adapters.rb]: 👍 Done

File: activerecord/lib/active_record/connection_adapters.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@ module ConnectionAdapters
       autoload :ReferenceDefinition
     end
 
-    autoload_at "active_record/connection_adapters/abstract/connection_pool" do
+    autoload_at "active_record/connection_adapters/abstract/connection_handler" do
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters.rb:28]: This can also be moved to `autoload_under`

Comment (gmcgibbon) [reply activerecord/lib/active_record/connection_adapters.rb]: Oops, you're right. Fixed!

**Review by rafaelfranca — APPROVED** [review]

**Comment by eileencodes** [comment]
This broke the tests in our internal adapter to throw a `NameError: uninitialized constant ActiveRecord::ConnectionAdapters::NullPool`. It doesn't affect the actual adapter in the app, just the tests for the adapter itself.

I can fix this by removing the autoload for `ConnectionPool` and adding a require to `ConnectionHandler` to load this file, but I don't quite get why this broke that and why that change is the one that fixes it. Adding a require in the adapter itself results in a circular require warning.

**Comment by rafaelfranca** [comment]
We can also register `NullPool` in the autoload list if it can be loaded without loading the `ConnectionPool`. It is indeed weird that this happened about this change.

**Comment by eileencodes** [comment]
I pushed a fix here https://github.com/rails/rails/pull/41384

I was going to write a test but because AR establishes a connection for all tests before any tests run I think that's loading the ConnectionPool. Our gem doesn't establish any default connections and just tests the adapter connection itself which is why we see it there.

**Comment by gmcgibbon** [comment]
👍 Thanks

Does it make sense to add an autoload for `AbstractPool` too? I assume it would have the same referencing issues.

**Comment by eileencodes** [comment]
I think AbstractPool doesn't have an issue because it's not initialized anywhere, but I'm not 100% on that.

**Pull request merged**